packet_forwarder/src/lora_pkt_fwd.c

```diff
@@ -3047,23 +3073,202 @@ static void gps_process_sync(void) {
+    uint8_t UBX_BYTES_AVAIL_REG = 0xFD;
+    uint8_t read_buf[2] = {0, 0};
+    struct i2c_msg rdwr_msgs[] = {
+        {
+            .addr = 0x42,
+            .len = 1,
+            .buf = &UBX_BYTES_AVAIL_REG,
+        },
+        {
+            .addr = 0x42,
+            .flags = I2C_M_RD,
+            .len = 2,
+            .buf = read_buf,
+        },
+    };
+    struct i2c_rdwr_ioctl_data rdwr_data = {
+        .msgs = rdwr_msgs,
+        .nmsgs = 2,
+    };
+    if (ioctl(gps_dev_fd, I2C_RDWR, &rdwr_data) < 0) {
+        *avail = 0;
+        perror("WARNING: failed to read GPS available data count");
+        return -1;
+    }
+    uint16_t avail_ = 0;
+    unpack_le_u16(&avail_, read_buf);
+    *avail = (size_t)avail_;
+    return 0;
+}
+
+static int i2c_gps_read(size_t n, uint8_t * dst) {
+    uint8_t UBX_DATA_REG = 0xFF;
+    struct i2c_msg rdwr_msgs[] = {
+        {
+            .addr = 0x42,
+            .len = 1,
+            .buf = &UBX_DATA_REG,
+        },
+        {
+            .addr = 0x42,
+            .flags = I2C_M_RD,
+            .len = n,
+            .buf = dst,
+        },
+    };
+    struct i2c_rdwr_ioctl_data rdwr_data = {
+        .msgs = rdwr_msgs,
+        .nmsgs = 2,
+    };
+    if (ioctl(gps_dev_fd, I2C_RDWR, &rdwr_data) < 0) {
+        perror("WARNING: failed to read from GPS buffer data");
+        return -1;
+    }
+    return 0;
+}
+
+#define DELAY_ON_I2C_GLITCH 50
+
+void thread_gps_i2c(void) {
+    size_t const volatile canary_h = 0xdeadfeed;
+    char read_buf[128]; /* buffer to receive GPS data */
+    size_t const volatile canary_l = 0xdeadfeed;
+    size_t wr_idx = 0;     /* pointer to end of chars in buffer */
+
+    /* variables for PPM pulse GPS synchronization */
+    enum gps_msg latest_msg; /* keep track of latest NMEA message parsed */
+
+    /* initialize some variables before loop */
+    memset(read_buf, 0, sizeof read_buf);
+
+    while (!exit_sig && !quit_sig) {
+        assert(canary_h == 0xdeadfeed);
+        assert(canary_l == 0xdeadfeed);
+
+        size_t rd_idx        = 0;
+        size_t frame_end_idx = 0;
+        size_t n_queued      = 0;
+
+        /* poll gps for count of buffer bytes */
+        while (i2c_gps_available(&n_queued)) {
+            MSG("WARNING: n_avail failed\n");
+            wait_ms(DELAY_ON_I2C_GLITCH);
+        }
+
+        if (n_queued == 0) {
+            wait_ms(200);
+            continue;
+        } else if (n_queued == 0x8000 || n_queued == 0x80) {
+            MSG("WARNING: n_queued value 0x%x appears to be glitchy\n");
+            wait_ms(DELAY_ON_I2C_GLITCH);
+            continue;
+        }
+
+        size_t const read_capacity = sizeof(read_buf) - wr_idx;
+        size_t const n_read        = n_queued < read_capacity ? n_queued : read_capacity;
+
+        if (i2c_gps_read(n_read, read_buf + wr_idx)) {
+            wait_ms(DELAY_ON_I2C_GLITCH);
+            continue;
+        }
+
+        wr_idx += n_read;
```

Comment:
If reading number of buffer bytes succeeds what are the chances we make it to here? Just trying to get a handle on which of the two calls is more likely to glitch: `i2c_gps_available` or `i2c_gps_read`. Maybe it's random so both calls are equally glitchy?

Reply:
> Maybe it's random so both calls are equally glitchy?

Yes, they can all glitch. The glitching is a result of the GPS clock-stretching, which the Broadcom I2C peripheral doesn't handle properly.

Reply:
> If reading number of buffer bytes succeeds what are the chances we make it to here? 

by here, do you mean line number 3191 or 3194? You can't get to either line if reading `n_available` is is not successful.

Reply:
I meant 3194. Since all of the I2C calls can glitch the question is irrelevant.